File: android_pos_app/app/src/main/java/com/pos2013/offline/utils/IdGenerator.kt
package com.pos2013.offline.utils

import android.content.Context
import java.util.UUID

/**
 * Generates unique identifiers for transactions and batches
 * REQUIRED for Protocol 201.3 compliance
 */
object IdGenerator {

    private const val PREFS_NAME = "pos_id_prefs"
    private const val KEY_LAST_STAN = "last_stan"

    /**
     * Generate unique local transaction ID
     * Format: txn_{uuid}
     * Used for idempotency - prevents duplicate transactions
     */
    fun generateLocalTxnId(): String {
        return "txn_${UUID.randomUUID().toString().replace("-", "")}"
    }

    /**
     * Generate unique batch ID
     * Format: batch_{timestamp}_{random}
     */
    fun generateBatchId(): String {
        val timestamp = System.currentTimeMillis()
        val random = (1000..9999).random()
        return "batch_${timestamp}_$random"
    }

    /**
     * Generate STAN (System Trace Audit Number)
     * 6-digit number (000001 - 999999) for transaction tracking
     * Uses DB-backed StanCounterRepository when available for atomic increments.
     * Falls back to SharedPreferences-based increment if DB is unavailable.
     */
    suspend fun generateStan(context: Context): String {
        return try {
            // Use DB-backed atomic counter (preferred)
            val repo = com.pos2013.offline.data.AppDatabase.getStanCounterRepository(context)
            repo.nextStanString()
        } catch (e: Exception) {
            // Fallback: prefs-based increment for backward compatibility
            val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            var lastStan = prefs.getInt(KEY_LAST_STAN, 0)

            // Increment and wrap at 999999
            lastStan = (lastStan + 1) % 1_000_000
            if (lastStan == 0) lastStan = 1 // Never use 000000

            // Save for next time
            prefs.edit().putInt(KEY_LAST_STAN, lastStan).apply()

            // Format as 6-digit with leading zeros
            lastStan.toString().padStart(6, '0')
        }
    }

    /**
     * Generate terminal ID
     * Format: TERM-{8 random chars}
     */
    fun generateTerminalId(): String {
        val random = UUID.randomUUID().toString()
            .replace("-", "")
            .take(8)
            .uppercase()
        return "TERM-$random"
    }

    /**
     * Generate merchant ID
     * Format: MRC-{4 digit random}
     */
    fun generateMerchantId(): String {
        val random = (1000..9999).random()
        return "MRC-$random"
    }
}

File: flutter_pos_2013/lib/core/utils/id_generator.dart
import 'dart:math';
import 'package:shared_preferences/shared_preferences.dart';
import 'package:uuid/uuid.dart';

/// ID Generator for Transactions, Batches, and STAN
class IdGenerator {
  static const String _stanKey = 'current_stan';
  static const int _maxStan = 999999;
  
  /// Generate a unique local transaction ID
  static String generateLocalTxnId() {
    final uuid = const Uuid().v4();
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    return 'TXN-$timestamp-${uuid.substring(0, 8)}';
  }

  /// Generate a 6-digit STAN (System Trace Audit Number)
  static Future<String> generateStan() async {
    final prefs = await SharedPreferences.getInstance();
    int currentStan = prefs.getInt(_stanKey) ?? 0;
    
    // Increment and wrap around
    currentStan = (currentStan + 1) % (_maxStan + 1);
    if (currentStan == 0) currentStan = 1;
    
    await prefs.setInt(_stanKey, currentStan);
    
    return currentStan.toString().padLeft(6, '0');
  }

  /// Generate a batch ID
  static String generateBatchId() {
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final random = Random().nextInt(9999).toString().padLeft(4, '0');
    return 'BATCH-$timestamp-$random';
  }

  /// Generate an order ID for offline orders
  static String generateOrderId() {
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    final random = Random().nextInt(999).toString().padLeft(3, '0');
    return 'ORD-$timestamp-$random';
  }

  /// Generate a settlement code (6-digit)
  static String generateSettlementCode() {
    final random = Random();
    final code = random.nextInt(1000000).toString().padLeft(6, '0');
    return code;
  }

  /// Generate a merchant reference
  static String generateMerchantReference() {
    final timestamp = DateTime.now().millisecondsSinceEpoch;
    return 'REF-$timestamp';
  }

  /// Generate a short code for redemption
  static String generateShortCode() {
    final random = Random();
    // Generate 6-digit code
    return random.nextInt(1000000).toString().padLeft(6, '0');
  }

  /// Generate a UUID v4
  static String generateUuid() {
    return const Uuid().v4();
  }

  /// Reset STAN counter (for testing)
  static Future<void> resetStan() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setInt(_stanKey, 0);
  }

  /// Get current STAN without incrementing
  static Future<String> getCurrentStan() async {
    final prefs = await SharedPreferences.getInstance();
    int currentStan = prefs.getInt(_stanKey) ?? 0;
    return currentStan.toString().padLeft(6, '0');
  }
}
